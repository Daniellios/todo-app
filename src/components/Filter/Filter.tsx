import React from "react"

const Filter = ({
  todoList,
  setTodoList,
  setStatus,
  filteredTodos,
  currentStatus,
}: IFilterProps) => {
  //Tasks Left Counter
  //По картинке примера немного непонятно как отображается количество оставшихся Todo, поэтому всегда считается количество именно незавершенных заданий
  // независимо от текущего фильтра
  const tasksLeft = (): string => {
    let msg = ""
    if (todoList) {
      let completedCount: ITask[] = todoList.filter((task) => !task.isCompleted)

      if (completedCount.length === 1) {
        msg = "1 item left"
        return msg
      } else if (completedCount.length > 1) {
        msg = `${completedCount.length} items left`
        return msg
      } else {
        msg = "Nothing to do :)"
        return msg
      }
    }
    return msg
  }

  const filterAll = (): void => {
    setStatus("all")
  }

  const filterActive = (): void => {
    setStatus("active")
  }

  const filterCompleted = (): void => {
    setStatus("completed")
  }

  const clearCompleted = (): void => {
    setTodoList(todoList.filter((task: ITask) => !task.isCompleted))
  }

  return (
    <div className="w-full flex justify-between  items-center">
      <span
        data-testid="items-counter"
        className="text-xs sm:text-lg text-paletteWhite w-[130px]"
      >
        {tasksLeft()}
      </span>
      <div className="flex gap-1 md:gap-4">
        <button
          onClick={filterAll}
          className={
            currentStatus === "all" ? "filter-button-active" : "filter-button"
          }
        >
          All
        </button>
        <button
          onClick={filterActive}
          className={
            currentStatus === "active"
              ? "filter-button-active"
              : "filter-button"
          }
        >
          Active
        </button>
        <button
          onClick={filterCompleted}
          className={
            currentStatus === "completed"
              ? "filter-button-active"
              : "filter-button"
          }
        >
          Completed
        </button>
      </div>

      <button onClick={clearCompleted} className="filter-button">
        Clear Completed
      </button>
    </div>
  )
}

export default Filter
