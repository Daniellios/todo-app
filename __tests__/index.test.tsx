import React from "react"
import { render, fireEvent, screen } from "@testing-library/react"
import "@testing-library/jest-dom"

import Home from "../src/pages"
import Filter from "../src/components/Filter/Filter"

const addTodo = (list: ITask[], item: ITask) => [...list, item]

const startTodoList = [
  { taskName: "Тестовое задание", isCompleted: true, id: 1 },
  { taskName: "Прекрасный Код", isCompleted: true, id: 2 },
  { taskName: "Покрытие тестами", isCompleted: true, id: 3 },
]

it("renders homepage correctly", () => {
  const { container } = render(<Home />)
  expect(container).toMatchSnapshot()
})

it("Adds new to do correctly", () => {
  const newTodo = { taskName: "Новое Задание", isCompleted: false, id: 4 }

  const expected = [
    { taskName: "Тестовое задание", isCompleted: true, id: 1 },
    { taskName: "Прекрасный Код", isCompleted: true, id: 2 },
    { taskName: "Покрытие тестами", isCompleted: true, id: 3 },
    { taskName: "Новое Задание", isCompleted: false, id: 4 },
  ]

  const result = addTodo(startTodoList, newTodo)
  expect(result).toEqual(expected)
})

test("if handles onclick ", () => {
  const onClick = jest.fn()
  render(<Home />)
  const buttonElement = screen.getByTestId("Add")
  buttonElement.addEventListener("click", onClick)
  fireEvent.click(buttonElement)
  expect(onClick).toHaveBeenCalledTimes(1)
})

test("if error is shown on empty input", () => {
  const onClick = jest.fn()
  render(<Home />)
  const inputElement = screen.getByPlaceholderText("Type your to do")
  const addTodoBtn = screen.getByText("Add To Do")
  addTodoBtn.addEventListener("click", onClick)
  fireEvent.click(addTodoBtn)
  expect(inputElement).toHaveClass("input-error")
})
