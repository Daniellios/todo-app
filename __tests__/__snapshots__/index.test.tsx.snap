// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders homepage correctly 1`] = `
<div>
  <main
    class="container mx-auto flex flex-col items-center justify-start h-screen p-4 "
  >
    <h1
      class="text-5xl md:text-[5rem] leading-normal font-thin text-pink-300/50 "
    >
      todos
    </h1>
    <div
      class="flex flex-col w-full min-w-[345px] gap-6 lg:w-[80%] xl:max-w-[800px] shadow-lg p-4 border-[1px] border-gray-500/10"
    >
      <div
        class="flex h-12 w-full border-[1px] border-gray-500/25 rounded"
      >
        <input
          class="h-full w-full px-4 rounded focus: border-none outline-none"
          placeholder="Type your to do"
          type="text"
          value=""
        />
        <button
          class="flex justify-center items-center text-gray-600 px-1 border-l-[1px] border-gray-500/25 text-center text-sm w-32 hover:cursor-pointer hover:bg-pink-300/40"
          data-testid="Add"
        >
          Add To Do
        </button>
      </div>
      <div
        class="w-full flex flex-col justify-center items-start gap-3 text-center"
      >
        <div
          class="w-full border-b-[1px] border-gray-500/25 "
        >
          <div
            class="flex justify-start items-center gap-8 py-2"
          >
            <button
              class="flex justify-center items-center w-8 h-8 rounded-full border-[1px] border-gray-400 hover:cursor-pointer hover:bg-slate-500/25"
            >
              <svg
                class="text-green-500"
                fill="currentColor"
                height="1.5rem"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 1024 1024"
                width="1.5rem"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M912 190h-69.9c-9.8 0-19.1 4.5-25.1 12.2L404.7 724.5 207 474a32 32 0 0 0-25.1-12.2H112c-6.7 0-10.4 7.7-6.3 12.9l273.9 347c12.8 16.2 37.4 16.2 50.3 0l488.4-618.9c4.1-5.1.4-12.8-6.3-12.8z"
                />
              </svg>
            </button>
            <h2
              class="text-xl text-gray-500/50 line-through"
            >
              Тестовое задание
            </h2>
          </div>
        </div>
        <div
          class="w-full border-b-[1px] border-gray-500/25 "
        >
          <div
            class="flex justify-start items-center gap-8 py-2"
          >
            <button
              class="flex justify-center items-center w-8 h-8 rounded-full border-[1px] border-gray-400 hover:cursor-pointer hover:bg-slate-500/25"
            >
              <svg
                class="text-green-500"
                fill="currentColor"
                height="1.5rem"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 1024 1024"
                width="1.5rem"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M912 190h-69.9c-9.8 0-19.1 4.5-25.1 12.2L404.7 724.5 207 474a32 32 0 0 0-25.1-12.2H112c-6.7 0-10.4 7.7-6.3 12.9l273.9 347c12.8 16.2 37.4 16.2 50.3 0l488.4-618.9c4.1-5.1.4-12.8-6.3-12.8z"
                />
              </svg>
            </button>
            <h2
              class="text-xl text-gray-500/50 line-through"
            >
              Прекрасный Код
            </h2>
          </div>
        </div>
        <div
          class="w-full border-b-[1px] border-gray-500/25 "
        >
          <div
            class="flex justify-start items-center gap-8 py-2"
          >
            <button
              class="flex justify-center items-center w-8 h-8 rounded-full border-[1px] border-gray-400 hover:cursor-pointer hover:bg-slate-500/25"
            >
              <svg
                class="text-green-500"
                fill="currentColor"
                height="1.5rem"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 1024 1024"
                width="1.5rem"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M912 190h-69.9c-9.8 0-19.1 4.5-25.1 12.2L404.7 724.5 207 474a32 32 0 0 0-25.1-12.2H112c-6.7 0-10.4 7.7-6.3 12.9l273.9 347c12.8 16.2 37.4 16.2 50.3 0l488.4-618.9c4.1-5.1.4-12.8-6.3-12.8z"
                />
              </svg>
            </button>
            <h2
              class="text-xl text-gray-500/50 line-through"
            >
              Покрытие тестами
            </h2>
          </div>
        </div>
      </div>
      <div
        class="w-full flex justify-between items-center"
      >
        <span
          class="text-xs sm:text-lg text-gray-500/70 w-[130px]"
          data-testid="items-counter"
        >
          Nothing to do :)
        </span>
        <div
          class="flex gap-1 md:gap-4"
        >
          <button
            class="filter-button-active"
          >
            All
          </button>
          <button
            class="filter-button"
          >
            Active
          </button>
          <button
            class="filter-button"
          >
            Completed
          </button>
        </div>
        <button
          class="filter-button"
        >
          Clear Completed
        </button>
      </div>
    </div>
  </main>
</div>
`;
